src/stackdriver-nozzle/stackdriver/metrics_buffer.go

```diff
@@ -16,12 +16,83 @@
 
 package stackdriver
 
-import "reflect"
+import (
+	"context"
+	"github.com/mitchellh/hashstructure"
+	"reflect"
+	"sync"
+	"time"
+)
 
 type MetricsBuffer interface {
 	PostMetric(*Metric)
 }
 
+type timerMetricsBuffer struct {
+	adapter MetricAdapter
+	errs    chan error
+	size    int
+	ticker  *time.Ticker
+	ctx     context.Context
+
+	metricsMu sync.Mutex // Guard metrics
+	metrics   metricsMap
+}
+
+func NewTimerMetricsBuffer(ctx context.Context, frequency time.Duration,
+	size int, adapter MetricAdapter) (MetricsBuffer, <-chan error) {
+	errs := make(chan error)
+	mb := &timerMetricsBuffer{
+		adapter: adapter,
+		errs:    errs,
+		metrics: make(map[string]*Metric),
+		size:    size,
+		ctx:     ctx,
+		ticker:  time.NewTicker(frequency),
+	}
+	mb.start()
+	return mb, errs
+}
+
+func (mb *timerMetricsBuffer) PostMetric(metric *Metric) {
+	mb.addMetric(metric)
+}
+
+func (mb *timerMetricsBuffer) addMetric(newMetric *Metric) {
+	k, _ := hashstructure.Hash(newMetric.Labels, nil)
+	mb.metricsMu.Lock()
+	defer mb.metricsMu.Unlock()
+	mb.metrics[newMetric.Name+string(k)] = newMetric
```

Comment:
this is the magic line right? where we overwrite an existing metric reading if it's in the buffer and we haven't flushed yet. 

pls comment here :)